zcash_primitives/src/primitives.rs

```diff
@@ -14,23 +15,72 @@ use crate::jubjub::{edwards, FixedGenerators, JubjubEngine, JubjubParams, PrimeO
 
 use blake2s_simd::Params as Blake2sParams;
 
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum AssetType {
+    Zcash,
+    Str4dBucks,
+}
+
+impl AssetType {
```

Comment:
This we should change a bit to be dynamic (e.g. take a string which becomes the tag)

Reply:
So, I did confuse one thing here. The u32 in the note is not the value base point (which is larger than 32 bits anyway) but rather is an (arbitrary) integer which is assigned to each type. The circuit then does not read this u32 directly, but instead is provided the value base in the witness.

I suppose there's some subjective decisions to make how to make this more dynamic and not have the asset list fixed at compile time. My first instinct is maybe to BLAKE hash the tag and let that be the u32, which would make writing from_note_plaintext complicated (it would have to lookup from a list) but at least to_note_plaintext would be simple.

Reply:
Right, that seems reasonable at first thought.